server/http4s-server/src/main/scala/sttp/tapir/server/http4s/EndpointToHttp4sServer.scala

```diff
@@ -1,70 +1,75 @@
+              }
 
-              case None => (values: DecodeInputsResult).pure[F]
-            }
-          case failure: DecodeInputsResult.Failure => (failure: DecodeInputsResult).pure[F]
-        }
-      }
-
-      def valueToResponse(value: Any): F[Response[F]] = {
-        val i = value.asInstanceOf[I]
-        se.logic(new CatsMonadError)(i)
-          .map {
-            case Right(result) => outputToResponse(ServerDefaults.StatusCodes.success, se.endpoint.output, result)
-            case Left(err)     => outputToResponse(ServerDefaults.StatusCodes.error, se.endpoint.errorOutput, err)
-          }
-          .flatTap { response => serverOptions.logRequestHandling.requestHandled(se.endpoint, response.status.code) }
-          .onError {
-            case e: Exception => serverOptions.logRequestHandling.logicException(se.endpoint, e)
+            case None => (values: DecodeInputsResult).pure[G]
           }
+        case failure: DecodeInputsResult.Failure => (failure: DecodeInputsResult).pure[G]
       }
+    }
 
-      OptionT(decodeBody(internal.DecodeInputs(se.endpoint.input, new Http4sDecodeInputsContext[F](req))).flatMap {
+    def valueToResponse(value: Any): G[Response[F]] = {
+      val i = value.asInstanceOf[I]
+      se.logic(new CatsMonadError[G])(i)
+        .map {
+          case Right(result) => outputToResponse(ServerDefaults.StatusCodes.success, se.endpoint.output, result)
+          case Left(err)     => outputToResponse(ServerDefaults.StatusCodes.error, se.endpoint.errorOutput, err)
+        }
+        .flatTap { response => t(serverOptions.logRequestHandling.requestHandled(se.endpoint, response.status.code)) }
+        .onError {
+          case e: Exception => t(serverOptions.logRequestHandling.logicException(se.endpoint, e))
+        }
+    }
+
+    Kleisli((req: Request[F]) =>
+      OptionT(decodeBody(req, internal.DecodeInputs(se.endpoint.input, new Http4sDecodeInputsContext[F](req))).flatMap {
         case values: DecodeInputsResult.Values =>
           InputValues(se.endpoint.input, values) match {
             case InputValuesResult.Value(params, _)        => valueToResponse(params.asAny).map(_.some)
-            case InputValuesResult.Failure(input, failure) => handleDecodeFailure(se.endpoint, input, failure)
+            case InputValuesResult.Failure(input, failure) => t(handleDecodeFailure(se.endpoint, input, failure))
           }
-        case DecodeInputsResult.Failure(input, failure) => handleDecodeFailure(se.endpoint, input, failure)
+        case DecodeInputsResult.Failure(input, failure) => t(handleDecodeFailure(se.endpoint, input, failure))
       })
-    }
-
-    service
+    )
   }
 
-  def toRoutes[I, E, O](serverEndpoints: List[ServerEndpoint[_, _, _, EntityBody[F], F]]): HttpRoutes[F] = {
-    NonEmptyList.fromList(serverEndpoints.map(se => toRoutes(se))) match {
+  def toHttp[I, E, O, G[_]: Sync](se: List[ServerEndpoint[_, _, _, EntityBody[F], G]])(t: F ~> G): Http[OptionT[G, *], F] =
```

Comment:
`I`, `E` and `O` are never used, so we can remove them.

Same parameters can also be removed removed from the corresponding `toRoutes` below, but since that would break source compatibility, we shouldn't touch it. It's also outside the scope of this PR.